feat: add 'alloydb-ai-nl' tool (#358)
googleapis/genai-toolbox: docs/en/resources/tools/alloydb-nla.md

---
title: "alloydb-ai-nl"
type: docs
weight: 1
description: > 

>   A "alloydb-ai-nl" tool leverages AlloyDB's AI functions to execute natural language questions against the database.

Author: https://cloud.google.com/alloydb/docs/natural-language-questions-overview this URL is 404 now and it caused the documentation build.

---

## About

A `alloydb-ai-nl` tool leverages AlloyDB's AI functions to execute natural language questions against the database. It allows users to query database information using natural language instead of SQL. It's compatible with the following sources:
- [alloydb-postgres](../sources/alloydb-pg.md)

The tool uses AlloyDB's natural language processing capabilities to interpret questions and convert them into appropriate SQL queries, which are then executed against the database. TODO: link to AlloyDB's documentation.

> 

Reviewer: ```suggestion

## Requirements

{{< notice tip >}} 
AlloyDB AI natural language is currently in gated public preview. For more information on availability and limitations, please see [AlloyDB AI natural language overview][https://cloud.google.com/alloydb/docs/natural-language-questions-overview].
{{< /notice >}}

To enable AlloyDB AI natural language for your AlloyDB cluster, please follow the steps listed in the [Generate SQL queries that answer natural language questions][alloydb-ai-gen-nl], including enabling the extension and configuring context for your application. 

[alloydb-ai-gen-nl]: https://cloud.google.com/alloydb/docs/alloydb/docs/ai/generate-queries-natural-language

```

> ## Fields
> 
> `nlConfig` is the name of the `nl_config` created in AlloyDB.
> 
> `nlConfigParameters` are the list of the parameters and values for the AlloyDB [PSV (parameterized secure views)](!https://cloud.google.com/alloydb/docs/ai/use-psvs#sanitize_queries_with_parameterized_secure_views).
> 
> When using this tool, all the PSV parameters should be from filled with values from an auth service or a bounded param. These parameters should not be visible to the LLM agent. Instead, the LLM will only see one argument when using this tool - `question`, with the description being "The natural language question to ask."

Reviewer: ```suggestion
## Configuration

### Specify Context

_Context_ includes any kind of information that can be used to answer an end user question. Context includes schema structure and relationships, summaries and descriptions of columns, column values and their semantics, and rules or statements of business logic specific to the application or domain.

Once you've follow the steps for configuring context, you can use the `context` field when configuring a `alloydb-ai-nl` tool. When this tool is invoked, the SQL will be generated and executed using this context. 

### Specifying Parameters to PSV's

[Parameterized Secure Views (PSVs)][alloydb-psv] are a feature unique to AlloyDB that allow you allow you to require one or more named parameter values passed to the view when querying it, somewhat like bind variables with ordinary database queries. You **must** supply all parameters required for all PSVs in the context. These parameters can be used with features like [Authenticated Parameters](../tools/#array-parameters) to provide secure access to queries generated using natural language.

[alloydb-psv]: https://cloud.google.com/alloydb/docs/ai/use-psvs#parameterized_secure_views

```

Reviewer: TODO: double check that relative link is correct

## Example

```yaml
tools:
  ask_questions:
    kind: alloydb-ai-nl
    source: my-alloydb-source
    description: "Ask questions to check information about flights"

>     nlConfig: "cymbal_air_nl_config"
>     nlConfigParameters:

Reviewer: the latest version of the documentation is referring to this as "context". I think we should update these fields to match.
```suggestion
    context: "cymbal_air_nl_config"
    contextParameters:
```

      - name: user_email
        type: string
        description: User ID of the logged in user.

>         authServices:
>           - name: my_google_service
>             field: email

Reviewer: ```suggestion
        # note: we strongly recommend using features like Authenticated or Bound parameters 
        # to prevent the LLM from seeing these params and specifying values it 
        # shouldn't in the tool input
        authServices:
          - name: my_google_service
            field: email
```
